pkg/service/model_builder_test.go

```diff
@@ -2930,53 +3006,3424 @@ func Test_defaultModelBuilderTask_Build(t *testing.T) {
+				subnetsResolver.EXPECT().ResolveViaDiscovery(gomock.Any(), gomock.Any()).Return(call.subnets, call.err)
+			}
+			for _, call := range tt.resolveViaNameOrIDSliceCalls {
+				subnetsResolver.EXPECT().ResolveViaNameOrIDSlice(gomock.Any(), gomock.Any(), gomock.Any()).Return(call.subnets, call.err)
+			}
+			featureGates := config.NewFeatureGates()
+			for key, value := range tt.featureGates {
+				if value {
+					featureGates.Enable(key)
+				} else {
+					featureGates.Disable(key)
+				}
+			}
+			annotationParser := annotations.NewSuffixAnnotationParser("service.beta.kubernetes.io")
+			trackingProvider := tracking.NewDefaultProvider("service.k8s.aws", "my-cluster")
+
+			elbv2TaggingManager := elbv2.NewMockTaggingManager(ctrl)
+			for _, call := range tt.listLoadBalancerCalls {
+				elbv2TaggingManager.EXPECT().ListLoadBalancers(gomock.Any(), gomock.Any()).Return(call.sdkLBs, call.err)
+			}
+			vpcInfoProvider := networking.NewMockVPCInfoProvider(ctrl)
+			for _, call := range tt.fetchVPCInfoCalls {
+				vpcInfoProvider.EXPECT().FetchVPCInfo(gomock.Any(), gomock.Any(), gomock.Any()).Return(call.wantVPCInfo, call.err).AnyTimes()
+			}
+			serviceUtils := NewServiceUtils(annotationParser, "service.k8s.aws/resources", "service.k8s.aws/nlb", featureGates)
+			defaultTargetType := tt.defaultTargetType
+			if defaultTargetType == "" {
+				defaultTargetType = "instance"
+			}
+			backendSGProvider := networking.NewMockBackendSGProvider(ctrl)
+			if tt.enableBackendSG {
+				backendSGProvider.EXPECT().Get(gomock.Any(), networking.ResourceType(networking.ResourceTypeService), gomock.Any()).Return(tt.backendSecurityGroup, nil).AnyTimes()
+				backendSGProvider.EXPECT().Release(gomock.Any(), networking.ResourceType(networking.ResourceTypeService), gomock.Any()).Return(nil).AnyTimes()
+			}
+			sgResolver := networking.NewMockSecurityGroupResolver(ctrl)
+			for _, call := range tt.resolveSGViaNameOrIDCall {
+				sgResolver.EXPECT().ResolveViaNameOrID(gomock.Any(), call.args).Return(call.want, call.err)
 			}
 			var enableIPTargetType bool
 			if tt.enableIPTargetType == nil {
 				enableIPTargetType = true
 			} else {
 				enableIPTargetType = *tt.enableIPTargetType
 			}
-			builder := NewDefaultModelBuilder(annotationParser, subnetsResolver, vpcInfoProvider, "vpc-xxx", trackingProvider, elbv2TaggingManager, featureGates,
-				"my-cluster", nil, nil, "ELBSecurityPolicy-2016-08", defaultTargetType, enableIPTargetType, serviceUtils)
+			builder := NewDefaultModelBuilder(annotationParser, subnetsResolver, vpcInfoProvider, "vpc-xxx", trackingProvider, elbv2TaggingManager, ec2Client, featureGates,
+				"my-cluster", nil, nil, "ELBSecurityPolicy-2016-08", defaultTargetType, enableIPTargetType, serviceUtils,
+				backendSGProvider, sgResolver, tt.enableBackendSG, tt.disableRestrictedSGRules)
 			ctx := context.Background()
-			stack, _, err := builder.Build(ctx, tt.svc)
+			stack, _, _, err := builder.Build(ctx, tt.svc)
 			if tt.wantError {
```

Comment:
`wantError` should be more specific: it should be a `string` and should `assert.EqualError()`